### PR TITLE
Ability to skip reconcilation by setting annotation

### LINKED ISSUE #509: Add an annotation that will prevent cluster restart
As a user I would like to decide which clusters will not go through restart (rolling update is a more accurate term) when the operator has a change that requires node restart

Adding an annotation that will tell the operator controller not to update this cluster now will help user control the timing of restarts

### PR DESCRIPTION
**Pull request by ChunyiLyu**
This closes #509

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed

## Summary Of Changes

- when 'rabbitmq.com/skipReconcile' is set to true on a rabbitmqcluster CR, Reconcile() will exit at the beginning of method
- Reconcile() logs and publishes event to warn users that it has skipped updating a CR
- tested only in integration level; I don't think this feature needs to be tested in system level since the operator uses annotations to indicate whether statefulSet or configmap is updated or not, and integration level covers this

### CODE REVIEW

File: controllers/rabbitmqcluster_controller.go [hunk 1]
```diff
@@ -101,6 +102,16 @@ func (r *RabbitmqClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 		return ctrl.Result{}, r.prepareForDeletion(ctx, rabbitmqCluster)
 	}
 
+	// exit if skip reconcile annotation is set to true
+	if v, ok := rabbitmqCluster.Annotations["rabbitmq.com/skipReconcile"]; ok && v == "true" {
```
Comment (ansd) [inline controllers/rabbitmqcluster_controller.go:106]: Let's use the constant defined above.

File: controllers/rabbitmqcluster_controller.go [hunk 2]
```diff
@@ -101,6 +102,16 @@ func (r *RabbitmqClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
...
+		logger.Info("Skip reconciling RabbitmqCluster",
+			"namespace", rabbitmqCluster.Namespace,
+			"name", rabbitmqCluster.Name)
+		r.Recorder.Event(rabbitmqCluster, corev1.EventTypeWarning,
+			"SkipReconcile", "annotation 'rabbitmq.com/skipReconcile' is set to true")
```
Comment (ansd) [inline controllers/rabbitmqcluster_controller.go:111]: Let's use the constant defined above.

**Review by ansd — COMMENTED** [review]
Should we add user documentation now or will we add docs once the feature is released?

File: controllers/rabbitmqcluster_controller.go [hunk 3]
```diff
@@ -101,6 +102,16 @@ func (r *RabbitmqClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
...
+		return ctrl.Result{}, err
```
Comment (ansd) [inline controllers/rabbitmqcluster_controller.go:112]: Shouldn't we update the status before skipping reconciliation?
Skipping reconciliation means we don't update the actual cluster to match the desired spec.
However, reporting the current status should still be done although reconciliation is skipped.
Imagine a pod dies. The `AllReplicasReadyCondition` condition should still be updated to be `false`.

**Review by ansd — CHANGES_REQUESTED** [review]

File: controllers/rabbitmqcluster_controller.go (hunk 3 above)
Comment (ansd) [reply controllers/rabbitmqcluster_controller.go:112]: Also, in the current implementation deletion is allowed even when reconciliation is skipped.
When reconciliation is skipped and the desired state is `Delete`, the operator shouldn't reconcile, i.e. it shouldn't call `prepareForDeletion`.

**Comment by ansd** [comment]
Based on the conversation with @mkuratczyk and @yaronp68:
* We think it makes sense to use a label instead of an annotation. Using a label has the advantage that a user can set it via `kubectl label` and easily query all instances where reconciliation is paused.
* We will add `kubectl rabbitmq pause-reconciliation` and `kubectl rabbitmq resume-reconciliation` commands. We should rename the label `rabbitmq.com/skipReconcile` to `rabbitmq.com/pauseReconciliation` to use the same wording.
* If reconciliation is paused, we should update the Status `NoWarnings` to be `false` with Reason `reconciliation paused`. This has the advantage that we also see the `lastTransitionTime`, i.e. since when reconciliation has been paused. We should leave the status `ReconcileSuccess` to be `true` (interpreted as "successfully skipped reconciliation").

**Comment by Zerpet** [comment]
> Based on the conversation with @mkuratczyk and @yaronp68:
> 
>     * We think it makes sense to use a label instead of an annotation. Using a label has the advantage that a user can set it via `kubectl label` and easily query all instances where reconciliation is paused.
> 
>     * We will add `kubectl rabbitmq pause-reconciliation` and `kubectl rabbitmq resume-reconciliation` commands. We should rename the label `rabbitmq.com/skipReconcile` to `rabbitmq.com/pauseReconciliation` to use the same wording.
> 
>     * If reconciliation is paused, we should update the Status `NoWarnings` to be `false` with Reason `reconciliation paused`. This has the advantage that we also see the `lastTransitionTime`, i.e. since when reconciliation has been paused. We should leave the status `ReconcileSuccess` to be `true` (interpreted as "successfully skipped reconciliation").

I think we should use an annotation because it has functional/behavioral implications. We can use `kubectl annotate` to easily annotate resources.

From the Kubernetes [docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/), annotations can be used for "directives from the end-user to the implementations to modify behavior or engage non-standard features".

Agreed that, by using annotations, we don't have the ability to query all the custom resources that are skipping reconciliation. If we want to include this feature, our plugin could also add a label, or the same label-value pair as the annotation-value pair, to the custom resource.

I believe that the use case to select all custom resources, from a given or all namespaces, that are skipping reconciliation should be uncommon because, as a user, you want the Operator to manage your RabbitMQ clusters.

File: controllers/rabbitmqcluster_controller.go (hunk 1 above)
Comment (huangjiasingle) [reply controllers/rabbitmqcluster_controller.go:106]: rabbitmqCluster.Annotations need to avoid go panic.

**Pull request closed without merging**